### DISCUSSION TITLE
data undefined

### DISCUSSION THREAD

**Discussion by xf15** [jspsych/jsPsych#1132] [Q&A] [answered]
i am trying to do something like this to save some randomly  generated id for subjects
```
var sample = {
  type: "html-keyboard-response",
  stimulus: function(data) {
     data.blah = jsPsych.randomization.randomID(8)
     return "copy to mturk" + data.blah
  } 
}
```

but i would get
`Uncaught TypeError: Cannot set property 'blah' of undefined`

i am referring to the use at https://github.com/jspsych/jsPsych/blob/master/examples/demo-flanker.html

i don't know why it doesn't work for me

**Comment by becky-gilbert** [comment, marked as answer]
The problem here is that, when you use a function for the `stimulus` parameter, that function won't be passed a data object, which is why you're getting the undefined error. I'm guessing you're getting this confused with the trial's `on_finish` function, which _is_ automatically passed an argument containing the data from that trial. Here are the [docs for the trial's on_finish function](https://www.jspsych.org/overview/callbacks/#on_finish-trial), which is an event-related function. And here are the [docs for dynamic parameters](https://www.jspsych.org/overview/trial/#dynamic-parameters), which is when functions are used for 'normal' parameters like `stimulus`.

Even though your stimulus function won't be automatically passed a data argument, you can still save things to the data inside that function - you just need to use one of [jsPsych's data functions](https://www.jspsych.org/core_library/jspsych-data/). One option is to use the [`jsPsych.data.addProperties` function](https://www.jspsych.org/core_library/jspsych-data/#jspsychdataaddproperties) to save the ID to every trial:
```javascript
var sample = {
  type: "html-keyboard-response",
  stimulus: function() {
    var blah = jsPsych.randomization.randomID(8);
    jsPsych.data.addProperties({rand_id: blah});
    return "copy to mturk" + blah
  } 
}
```
There's also [this example](https://www.jspsych.org/overview/data/#adding-data-to-all-trials) that shows how to generate an ID and save it to the data when the page first loads (rather than when a trial first starts, as in your approach).

**Reply by xf15** [reply]
ah, thank you @becky-gilbert
